zebra/zebra_mpls.c

```diff
@@ -2999,11 +3003,26 @@ int zebra_mpls_write_label_block_config(struct vty *vty, struct zebra_vrf *zvrf)
 /*
  * Called when VRF becomes inactive, cleans up information but keeps
  * the table itself.
- * NOTE: Currently supported only for default VRF.
  */
 void zebra_mpls_cleanup_tables(struct zebra_vrf *zvrf)
 {
-	hash_iterate(zvrf->lsp_table, lsp_uninstall_from_kernel, NULL);
+	if (zvrf_id(zvrf) == VRF_DEFAULT)
+		hash_iterate(zvrf->lsp_table, lsp_uninstall_from_kernel, NULL);
+	else {
+		/*
+		 * For other vrfs, we try to remove associated LSPs; we locate
+		 * the LSPs in the default vrf.
+		 */
+		struct zebra_vrf *def_zvrf;
+		afi_t id;
+
+		def_zvrf = zebra_vrf_lookup_by_id(VRF_DEFAULT);
+
+		for (id = AFI_IP; id < AFI_MAX; id++) {
```

Comment:
Why the delete code is different for default VRF and non default VRF in this case ? There are multiple differences.
1. I see that lsp_uninstall is deleting each NHLFE & ILM and only for dynamic LSPs. If non default VRF has static LSPs, how does these gets cleaned up ?
2. lsp_uninstall is not calling lsp_uninstall_from_kernel. Is it required to clean the LSPs of non-default VRF from kernel ? If so, how ?
3.

Reply:
Let's see:
the delete code is different because all of the lsps live in the default vrf's `lsp_table`. that's the initial issue - the non-default vrfs' have empty tables, so their data structs get left behind.

1) I ... don't know the answer to that, personally. maybe someone would like to track down whether that's an issue? the dynamic lsps are definitely an issue.
2) I think that the kernel/lfib uninstall happens once the lsp datastructure gets examined in lsp_process() after it's been marked 'uninstalled' here

Reply:
Did you test this particular scenario and checked if these LSPs are cleaned from kernel ?

For #2, we should definitely understand, which code actually cleans the LSPs from Kernel in this case. 

I will review the new changes as well.

Reply:
New changes to check whether default VRF is already destroyed or not looks good.

Reply:
yes, you can reproduce this issue in the topotest called `bgp_l3vpn_to_bgp_vrf`. just un-comment the cli "CheckFunc" line in test_check_linux_mpls(). try using router r1 - remove the vrf with `r1 ip link delete r1-cust1`, then try "r1 vtysh", and "show mpls table":  you'll see that zebra crashes: the lsp is still in the database, but the associated interface is NULL. with the fix, you'll see that the LSP is gone, both from zebra and from the kernel.